### ISSUE TITLE
Inconsistent `request.form()` return values for multipart vs urlencode

### ISSUE THREAD

**Issue by giovanism** [RobertoPrevato/BlackSheep#21]
I think the urlencode return value is the best presentation. So, even if the content-type is multipart, request.form() should also return dictionary.

**Comment by RobertoPrevato** [comment]
Hi @giovanism, thanks for your feedback.
I indeed was wondering what's the best API to handle both cases, so I welcome discussion about this topic.
 
Both `multipart/form-data` and `application/x-www-form-urlencoded` can contain duplicate keys. Duplicated keys for `multipart/form-data` seems to be more common: for example, when several files are uploaded with the same field name (an HTML file input with multiple).

To achieve consistency, the more logical solution would be to always return a dictionary of lists, but it looks a bit ugly when you have a single value for a key (arguably the most common scenario).

For example,
```python
{
    "name": ["Bartleby"],
    "some_duplicated_key": ["One", "Two", "Three"]
}
```

Besides, it's also inefficient to instantiate a `list` to store a single value. I considered to instantiate the list only when a given key is duplicated and return such objects:
```python
{
    "name": "Bartleby",   # <-- NB: since the key has a single value, it's not in a list
    "some_duplicated_key": ["One", "Two", "Three"]
}
```

But it might look surprising. How would you recommend to handle duplicated keys?

**Comment by giovanism** [comment]
Sorry, I wasn't specific enough. What I originally complained about was when the content_type is `multipart/form-data`, the returned value of `request.form()` is a list of `FormData` objects. But when the content_type is `application/x-www-form-urlencoded`, the returned value is a dict with field name as key and list of values as its values.

I think it would be more preferable to have more consistent return type. `request.form()` will return dict of form field names and values regardless of content_type. For convenience sake, this will assume single field value as is and duplicated field values as list. Then make another method/default argument to for field values as list in case anyone needs to get that result.

Then another method, for example `request.files()`, for file uploads.

**Comment by RobertoPrevato** [comment]
Sorry for replying late; you were specific and I understood what you meant. I agree and I am going to make the code API consistent for `request.form()`. My previous considerations are still actual; since in HTTP it's licit to have several fields (not only files) with the same name in multipart form data, I don't like to assume this will happen only for file uploads. I will implement a solution to handle several values almost transparently for the end user.

I'm currently working on several things, so it will take a few days more.

**Comment by giovanism** [comment]
Great! Hacktoberfest is near and I'm interested to work on this issue. So, can you let me work on this?

**Comment by RobertoPrevato** [comment]
I would like to take care of this task, since it's about existing code and I already dedicated a good deal of time thinking about this topic. But I would be happy to review and accept PRs, if you like to contribute, on one of these (potentially more interesting) things:

* support for [partial requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests) , like: server returning Accept-Ranges header and handling Ranges requests when serving static files
* support for gzip, deflate, br (both incoming and outgoing)

**Comment by RobertoPrevato** [comment]
@giovanism, I followed your recommendation and made consistent, as much as possible, the return value of `form()` method. Fixed at https://github.com/RobertoPrevato/BlackSheep/commit/3d9acdff4fb1d481225ceacfc1b6a3a095ac8ef5

* already published in [PyPi, with version 0.1.9](https://pypi.org/project/blacksheep/)

**Bonus**:
1. added `FromForm` binder, so now input from form payload can be parsed automatically into a model, by type annotation, for example:

```python
from blacksheep.server.bindings import FromForm

class Item:
    def __init__(self, a, b, c):
        self.a = a
        self.b = b
        self.c = int(c)


@app.router.post('/')
async def home(request, item: FromForm(Item)):
    assert item is not None
    assert item.a == 'Hello'
    assert item.b == 'World'
    assert item.c == 10
```

2. Added support for request handlers as class methods, [using Controllers](https://github.com/RobertoPrevato/BlackSheep/blob/master/tests/test_controllers.py)
3. Added PyCharm Stubs for all Cython classes, to improve developer's experience with PyCharm
4. Improved the handling of form data in general, correcting several bugs

As I said, I will welcome PRs on the topics listed above, or work on proof of concepts in general. I am working on project templates.

Thanks for the heads up!
